### PR TITLE
Add qubes.VMExec call, for running a single command

### PR DESCRIPTION
**Pull request by pwmarcz**
https://github.com/QubesOS/qubes-issues/issues/4850

### CODE REVIEW

File: qubesagent/vmexec.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    pass
+
+
+def decode_part(part):
+    if not re.match(r'^[a-zA-Z0-9._-]*$', part):
+        raise DecodeError('illegal characters found')
+
+    part = part.encode('ascii')
+
+    part = part.replace(b'--', b'/')
```
Comment (marmarek) [inline qubesagent/vmexec.py]: This could use a comment, especially why it's safe.

File: qubesagent/test_vmexec.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    def test_01_decode_empty(self):
+        self.assertEqual(decode('echo+'), [b'echo', b''])
+
+    def test_02_decode_escaping(self):
+        self.assertEqual(decode('echo+Hello-20world'),
+                         [b'echo', b'Hello world'])
+        self.assertEqual(decode('-0a-0D'),
+                         [b'\n\r'])
+        self.assertEqual(decode('ls+--la'),
+                         [b'ls', b'-la'])
```
Comment (marmarek) [inline qubesagent/test_vmexec.py]: I'd add also a test like `ls+---61`

File: qubesagent/test_vmexec.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    def test_03_decode_errors(self):
+        with self.assertRaises(DecodeError):
+            decode('illegal/slash')
+        with self.assertRaises(DecodeError):
+            decode('illegal-singledash')
```
Comment (marmarek) [inline qubesagent/test_vmexec.py]: And here a test for incomplete encoded byte (at the end, and/or in the middle)

**Comment by marmarek** [comment]
Also, besides commit messages, please also add issue reference(s) to the pull request description. Github the nicely list related PRs in the issue, including their state.

File: qubesagent/vmexec.py (hunk 1 above)
Comment (pwmarcz) [reply qubesagent/vmexec.py]: Right, it's actually not safe :( Things like `-2Fbin-2Fls` get converted to `/bin/ls` then replaced to `-bin-ls`. I found a better way.

**Comment by pwmarcz** [comment]
Updated:
- changed the approach to less hacky, fixing the `/` case
- permitted only capital letters in hex
- stronger tests (slashes, three/four dashes, small letters, incomplete byte)

File: qubesagent/vmexec.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    pass
+
+
+def decode_part(part):
+    if not re.match(r'^[a-zA-Z0-9._-]*$', part):
+        raise DecodeError('illegal characters found')
+
+    part = part.encode('ascii')
+
+    escape_re = br'--|-([A-F0-9]{2})'
```
Comment (marmarek) [inline qubesagent/vmexec.py]: re.compile?

Comment (pwmarcz) [reply qubesagent/vmexec.py]: Added.

**Pull request merged**